ion-schema/src/type_reference.rs

```diff
@@ -0,0 +1,72 @@
+use crate::ion_path::IonPath;
+use crate::isl::isl_type_reference::IslTypeRefModifier;
+use crate::result::ValidationResult;
+use crate::system::{TypeId, TypeStore};
+use crate::types::TypeValidator;
+use crate::IonSchemaElement;
+use ion_rs::IonType;
+
+/// Provides an internal representation of a schema type reference.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TypeReference {
```

Comment:
Adds a new implementation for type references which can be used to store type references in constraints. This also contains modifier related information (e.g. `nullable`, `$null_or`).